icarusalg/Utilities/TrackTimeInterval.cxx

```diff
@@ -0,0 +1,318 @@
+auto lar::util::TrackTimeInterval::timeRange(recob::Hit const& hit) const
+  -> TimeRange
+{
+  return timeRange(hit.PeakTime(), hit.WireID()); // that was easy...
+}
+
+
+// -----------------------------------------------------------------------------
+auto lar::util::TrackTimeInterval::timeRange(
+  detinfo::timescales::TPCelectronics_tick_d TPCtick,
+  geo::PlaneID const& planeID /* = ... */
+) const -> TimeRange {
+  /*
+   * The time of arrival of some charge deposited at time t0 and at distance dX
+   * from an anode plane is:
+   *   t(P) = t0 + dt(D) = t0 + dX / v
+   * where dt(D) is the time it takes the charge to drift to the plane and v is
+   * the drift velocity, assumed uniform.
+   * That time of arrival is directly converted into a tick R(P) by the TPC
+   * readout electronics:
+   *   R(P) = R(ref) + (t(P) - t(ref)) / T
+   * where R(ref) is by construction the readout tick measured at the reference
+   * time t(ref), t(ref) is on the same time scale as t0 and T is the tick
+   * period (400 ns in ICARUS, 500 ns in MicroBooNE and others).
+   * R(P) is also the tick reconstructed by the hit reconstruction,
+   * `recob::Hit::PeakTime()`.
+   * 
+   * In this function we are interested in t0:
+   *
+   *     t(P) = t(ref) + T (R(P) - R(ref))
+   *     t0 = t(P) - dX / v = t(ref) + T (R(P) - R(ref)) - dX / v
+   *
+   * We don't know what dX is and the point of this function is to return the
+   * allowed interval of t0, which is covered by all the values of dX between
+   * `0` (on the anode plane) and `D` (distance of the cathode from the anode
+   * plane):
+   * 
+   *     t0(max) = t(ref) + T (R(P) - R(ref))
+   *     t0 in [ t0(max) - D / v ; t0(max) ]
+   *
+   * t(ref) is the time of the trigger, so if we express t0 relative to the
+   * trigger time we can omit it. The electronics time, on the other end,
+   * does not refer to the trigger time.
+   * `DetectorTimings` understands the time scale of R(P) (TPC electronics tick)
+   * and can convert it to other scales, e.g. the "standard" electronics scale,
+   * and give directly t0(max) in those other scales.
+   */
+  
+  detinfo::timescales::electronics_time const time
+    = fDetTimings.toElectronicsTime(TPCtick);
+  
+  TimeLimits_t const& planeTimeLimits = fGeomCache.limits[planeID];
+  microseconds const driftTime{
+    planeTimeLimits.driftDistance.convertInto<centimeters>().value()
+    / fDriftVelocity
+    };
+  return { time - driftTime, time };
+  
+} // lar::util::TrackTimeInterval::timeRange(tick)
+
+
+// -----------------------------------------------------------------------------
+lar::util::TrackTimeInterval::TrackTimeInterval(
+  GeometryCache_t geomCache,
+  detinfo::DetectorPropertiesData detProp,
+  detinfo::DetectorTimings detTimings
+)
+  : fDetProp{ std::move(detProp) }
+  , fDetTimings{ std::move(detTimings) }
+  , fDriftVelocity{ fDetProp.DriftVelocity() }
+  , fGeomCache{ std::move(geomCache) }
```

Comment:
Silly C++ question time: Is the Maker class passing its `GeometryCache_t` as a copy to the constructor? Curious because of the `move` here given that one would want to use the Maker and its `GeometryCache_t` on the subsequent event as well.

Reply:
Not silly at all. The constructor is declared as `TrackTimeInterval(GeometryCache_t geomCache, detinfo::DetectorPropertiesData detProp, detinfo::DetectorTimings detTimings)`. When you call such function, e.g. `TrackTimeInterval{ cache, dProp, dTimes }`, `cache` is _copied_ into `geomCache` parameter, `dProp` into `detProp` etc. The `move()` tries to avoid a second copy from e.g. `cache` to `fGeomCache`.
This is a deliberate pattern. Another is to pass by reference and copy in the constructor (`TrackTimeInterval(GeometryCache_t const& geomCache, detinfo::DetectorPropertiesData const& detProp, detinfo::DetectorTimings const& detTimings)` and `fGeomCache{ geomCache }`). The only drawback I see in this pattern, and I am not sure if there is a C++ trick to avoid it, is that if I put a temporary object in the call, my solution will directly bind that temporary to `geomCache` (no copy) and then the move will also not copy. Conversely, in the by-reference approach the temporary will be bound to a constant reference (C++ specifically provides for that, still no copy), then there will be a copy in the constructor (`fGeomCache{ geomCache }`).

Reply:
Thanks for the explanation!